packages/core/src/typeschema/validation.ts

```diff
@@ -360,3 +388,67 @@ function checkObjectForNull(obj: Record<string, unknown>, path: string, issues:
     }
   }
 }
+
+function matchesSpecifiedValue(value: TypedValue | TypedValue[], element: ElementValidator): boolean {
+  if (element.pattern && !deepIncludes(value, element.pattern)) {
+    return false;
+  } else if (element.fixed && !deepEquals(value, element.fixed)) {
+    return false;
+  }
+  return true;
+}
+
+function matchDiscriminant(
+  value: TypedValue | TypedValue[] | undefined,
+  discriminator: SliceDiscriminator,
+  slice: SliceDefinition
+): boolean {
+  const element = slice.fields[discriminator.path];
+  switch (discriminator.type) {
+    case 'value':
+    case 'pattern':
+      if (!element) {
+        throw new OperationOutcomeError(
+          serverError(new Error(`Failed to match slicing discriminator at ${discriminator.path}`))
+        );
+      } else if (!value) {
+        return false;
+      } else if (matchesSpecifiedValue(value, element)) {
+        return true;
+      }
+      break;
+    case 'exists':
+      if (slice.min === 0 && slice.max === 0 && !value) {
+        return true;
+      } else if (slice.min > 0 && value) {
+        return true;
+      }
+      break;
+    case 'type':
+      if (!value) {
+        return false;
+      } else if (Array.isArray(value)) {
+        return value.some((v) => element.type.some((t) => t.code === v.type));
+      } else {
+        return element.type.some((t) => t.code === value.type);
+      }
+  }
+  // Default to no match
+  return false;
+}
+
+function checkSliceElement(value: TypedValue, slicingRules: SlicingRules): string | undefined {
+  nextSlice: for (const slice of slicingRules.slices) {
+    for (const discriminator of slicingRules.discriminator) {
+      const discrimValues = arrayify(getNestedProperty(value, discriminator.path));
+      if (!discrimValues) {
+        return undefined;
+      }
+      if (!discrimValues.some((v) => matchDiscriminant(v, discriminator, slice))) {
+        continue nextSlice;
```

Comment:
If you want to avoid the `nextSlice:` label, I think you could replace this with a simple `break;`, no?

Reply:
Changing this to a `break` would alter the semantics of the loop due to the `return` below, but I can take another crack at refactoring to remove the label 👍